static/app/components/nav/utils.tsx

```diff
@@ -0,0 +1,170 @@
+import type {ComponentProps} from 'react';
+import type {LocationDescriptor} from 'history';
+
+import type Feature from 'sentry/components/acl/feature';
+import {isItemActive} from 'sentry/components/sidebar/sidebarItem';
+import {SIDEBAR_NAVIGATION_SOURCE} from 'sentry/components/sidebar/utils';
+import normalizeUrl from 'sentry/utils/url/normalizeUrl';
+import type {useLocation} from 'sentry/utils/useLocation';
+
+/** NavItem is the base class for both SidebarItem and SubmenuItem  */
+interface NavItem {
+  /** User-facing item label, surfaced in the UI. Should be translated! */
+  label: string;
+  /** Optionally, props which should be passed to a wrapping `<Feature>` guard */
+  feature?: Omit<ComponentProps<typeof Feature>, 'children'>;
+}
+
+/** NavItems are displayed in either `main` and `footer` sections */
+export interface NavItemLayout<Item extends NavItem> {
+  main: Item[];
+  footer?: Item[];
+}
```

Comment:
I opted to keep the generic but update it to use the suggested union in 4e5864ddcbf. That seemed cleaner because it won't require guards when accessing individual elements.